src/cryptonote_core/service_node_list.cpp

```diff
@@ -312,7 +320,114 @@ namespace service_nodes
 
 
     m_rollback_events.push_back(std::unique_ptr<rollback_event>(new rollback_change(block_height, key, iter->second)));
+    uint64_t swarm_id = iter->second.swarm_id;
     m_service_nodes_infos.erase(iter);
+    check_and_process_swarm_decomission(swarm_id);
+  }
+
+  void service_node_list::check_and_process_swarm_decomission(uint64_t swarm_id)
+  {
+    size_t count = 0;
+    for (const auto& entry : m_service_nodes_infos)
+      if (entry.second.swarm_id == swarm_id)
+        count++;
+    if (count == MIN_SWARM_SIZE - 1)
+    {
+      // XXX XXX XXX XXX
+      // All nodes should fetch data from this swarm at this point only.
+      // The registered nodes will eventually disappear and after this point,
+      // it is already considered gone. It only exists to retrieve data from.
+      //
+      // XXX XXX XXX XXX
+      //         internship optimization hardfork idea:
+      //         when nodes have been decomissioned for more than 10 blocks,
+      //         move the nodes into new swarms immediately.
+    }
+  }
+
+  uint64_t service_node_list::get_new_node_swarm_id(uint64_t block_height, uint32_t index) const
+  {
+    // XXX XXX XXX improvement proposal: have the swarm id be "remembered" from
+    // the last time, if the new node was already registered before. This would
+    // have huge performance gains preventing nodes from having to resync every
+    // time they rejoin the network.
+
+    crypto::hash hash = m_blockchain.get_block_id_by_height(block_height);
+    uint64_t seed = 0;
+    std::memcpy(&seed, hash.data, sizeof(seed));
+    seed += index;
+    std::mt19937_64 mersenne_twister(seed);
```

Comment:
Can move this seeding work to below the early out case `if (swarm_ids.empty())`

Reply:
lol ok